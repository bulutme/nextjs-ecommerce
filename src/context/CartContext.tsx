import React, {
  createContext,
  useContext,
  ReactNode,
  useEffect,
  useState,
} from "react";
import { ProductItemProps } from "@/app/types/Product/types";
import useLocalStorage from "@/hooks/useLocalStorage";

type CartItem = {
  product: ProductItemProps;
  count: number;
};

type GroupedCart = {
  brand: string;
  products: CartItem[];
};

type CartContextType = {
  cart: GroupedCart[];
  addToCart: (product: ProductItemProps) => void;
  removeProduct: (productId: number) => void;
  increaseCount: (productId: number) => void;
  decreaseCount: (productId: number) => void;
  totalItemCount: number;
  totalPrice: string;
  isLoading: boolean;
};

const CartContext = createContext<CartContextType | undefined>(undefined);

const CartProvider = ({ children }: { children: ReactNode }) => {
  const [cart, setCart] = useLocalStorage<GroupedCart[]>("cart", []);
  const [isLoading, setIsLoading] = useState<boolean>(true);

  useEffect(() => {
    if (typeof window !== "undefined") {
      setIsLoading(true);
      setCart(JSON.parse(localStorage.getItem("cart") as string));
      setIsLoading(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const totalItemCount = cart.reduce(
    (total, group) =>
      total +
      group.products.reduce((groupTotal, item) => groupTotal + item.count, 0),
    0
  );

  const totalPrice = cart.reduce(
    (total, group) =>
      total +
      group.products.reduce(
        (groupTotal, item) => groupTotal + item.count * item.product.price,
        0
      ),
    0
  );

  const formattedTotalPrice = totalPrice.toFixed(2);

  const addToCart = (product: ProductItemProps) => {
    const existingGroupIndex = cart.findIndex(
      (group) => group.brand === product.brand
    );

    if (existingGroupIndex !== -1) {
      const existingProductIndex = cart[existingGroupIndex].products.findIndex(
        (item) => item.product.id === product.id
      );

      const newCart = [...cart];

      if (existingProductIndex !== -1) {
        const cartId =
          newCart[existingGroupIndex].products[existingProductIndex].product.id;
        increaseCount(cartId);
      } else {
        newCart[existingGroupIndex].products.push({
          product,
          count: 1,
        });
        setCart(newCart);
      }
    } else {
      setCart((prevCart) => [
        ...prevCart,
        {
          brand: product.brand,
          products: [
            {
              product,
              count: 1,
            },
          ],
        },
      ]);
    }
  };

  const removeProduct = (productId: number) => {
    const newCart = cart.map((group) => ({
      ...group,
      products: group.products.filter((item) => item.product.id !== productId),
    }));

    // Remove empty groups
    const updatedCart = newCart.filter((group) => group.products.length > 0);
    setCart(updatedCart);
  };

  const increaseCount = (productId: number) => {
    const newCart = cart.map((group) => {
      return {
        ...group,
        products: group.products.map((item) =>
          item.product.id === productId
            ? { ...item, count: item.count + 1 }
            : item
        ),
      };
    });

    setCart(newCart);
  };

  const decreaseCount = (productId: number) => {
    const newCart = cart.map((group) => ({
      ...group,
      products: group.products.map((item) =>
        item.product.id === productId
          ? { ...item, count: Math.max(1, item.count - 1) }
          : item
      ),
    }));

    setCart(newCart);
  };

  const contextValue: CartContextType = {
    cart,
    totalItemCount,
    totalPrice: formattedTotalPrice,
    addToCart,
    removeProduct,
    increaseCount,
    decreaseCount,
    isLoading,
  };

  return (
    <CartContext.Provider value={contextValue}>{children}</CartContext.Provider>
  );
};

const useCart = () => {
  const context = useContext(CartContext);

  if (!context) {
    throw new Error("useCart must be used within a CartProvider");
  }
  return context;
};

export { CartProvider, useCart };
