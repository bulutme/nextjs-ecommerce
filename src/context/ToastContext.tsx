import { createContext, FC, ReactNode, useContext, useState } from "react";

export type ToastType = "success" | "error" | "exit";

type ToastData = {
  id: number;
  type: ToastType;
  message: string;
};

type ToastContextType = {
  toasts: ToastData[];
  addToast: (type: ToastType, message: string) => void;
  removeToast: (id: number) => void;
};

const MAX_TOASTS = 5;

const ToastContext = createContext<ToastContextType | undefined>(undefined);

export const ToastProvider: FC<{ children: ReactNode }> = ({ children }) => {
  const [toasts, setToasts] = useState<ToastData[]>([]);

  // function to add toast using type and message
  const addToast = (type: ToastType, message: string) => {
    const id = new Date().getTime();
    const newToast: ToastData = { id, type, message };

    // remove the latest one if toast item more than 5
    setToasts((prevToasts) => [...prevToasts.slice(-MAX_TOASTS + 1), newToast]);

    if (type !== "exit") {
      setTimeout(() => {
        removeToast(id);
      }, 3000);
    }
  };

  // function to remove toast
  const removeToast = (id: number) => {
    setToasts((prevToasts) => prevToasts.filter((toast) => toast.id !== id));
  };

  const contextValue: ToastContextType = {
    toasts,
    addToast,
    removeToast,
  };

  return (
    <ToastContext.Provider value={contextValue}>
      {children}
    </ToastContext.Provider>
  );
};

// custom hook to use toast context
export const useToast = () => {
  const context = useContext(ToastContext);
  if (!context) {
    throw new Error("useToast must be used within a ToastProvider");
  }
  return context;
};
