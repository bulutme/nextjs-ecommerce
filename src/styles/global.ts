"use client";

import { createGlobalStyle } from "styled-components";

const GlobalStyles = createGlobalStyle`
  * {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
  }

  html {
    font-size: 16px;
  }

  html, body, #__next {
    height: 100%;
  }

  a {
    text-decoration: none;
    width: fit-content;
  }

  ul {
    list-style-type: none
  }
`;

export default GlobalStyles;
