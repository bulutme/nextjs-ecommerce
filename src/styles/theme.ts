export const theme = {
  colors: {
    black: "#000",
    white: "#fff",
    gray: "#808080",
    grayLight: "#d3d3d3",
    border: "#e2e8f0",
    primary: "#3bb77e",
    primaryDark: "#2b8a5e",
    error: "#dc2626",
    info: "#0891b2",
    success: "#16a34a",
  },
  screens: {
    "2xs": "400px",
    xs: "480px",
    sm: "576px",
    md: "768px",
    lg: "992px",
    xl: "1200px",
    "2xl": "1440px",
    "3xl": "1560px",
  },
  fontWeight: {
    normal: 400,
    medium: 500,
    semibold: 600,
    bold: 700,
  },
};
