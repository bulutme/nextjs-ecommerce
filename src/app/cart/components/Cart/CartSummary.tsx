import styled from "styled-components";
import Button from "../../../../components/Button";
import { useCart } from "@/context/CartContext";
import { FC } from "react";

const SummaryTitle = styled.h3`
  font-size: 1.25rem;
  margin-bottom: 1rem;
  font-weight: ${({ theme }) => theme.fontWeight.medium};

  @media (min-width: ${({ theme }) => theme.screens.md}) {
    text-align: center;
  }
`;

const CartSummaryContainer = styled.div`
  background-color: ${({ theme }) => theme.colors.white};
  border: 0;
  padding: 8px 12px;
  margin-bottom: 1rem;

  @media (min-width: ${({ theme }) => theme.screens.md}) {
    border: 1px solid ${({ theme }) => theme.colors.border};
    border-radius: 8px;
  }
`;

const CartItemsWrapper = styled.div`
  display: flex;
  flex-direction: column;

  & > div:not(:last-child) {
    border-bottom: 1px solid ${({ theme }) => theme.colors.border};
  }
`;

const CartItemRow = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  font-size: 0.875rem;
  padding: 8px 0;
  margin-top: 1px;
`;

const PriceTitle = styled.span`
  font-size: 0.875rem;
`;

const Price = styled.span`
  font-weight: ${({ theme }) => theme.fontWeight.semibold};
  font-size: 1rem;
  color: ${({ theme }) => theme.colors.black};
`;

const TotalRow = styled(CartItemRow)`
  margin-top: 8px;
`;

const TotalPrice = styled(Price)`
  font-size: 1.25rem;
  color: ${({ theme }) => theme.colors.primary};
`;

const CartSummary: FC = () => {
  const { totalPrice } = useCart();

  return (
    <>
      <CartSummaryContainer>
        <SummaryTitle>Cart Summary</SummaryTitle>
        <CartItemsWrapper>
          <CartItemRow>
            <PriceTitle>Subtotal</PriceTitle>
            <Price>${totalPrice}</Price>
          </CartItemRow>
          <CartItemRow>
            <PriceTitle>Shipping</PriceTitle>
            <Price>Free</Price>
          </CartItemRow>
          <TotalRow>
            <PriceTitle>Total</PriceTitle>
            <TotalPrice>${totalPrice}</TotalPrice>
          </TotalRow>
        </CartItemsWrapper>
      </CartSummaryContainer>
      <Button $fullwidth={true} $variant="rounded" color="primary">
        Proceed to checkout
      </Button>
    </>
  );
};

export default CartSummary;
