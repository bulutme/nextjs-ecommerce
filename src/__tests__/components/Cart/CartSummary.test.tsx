import React from "react";
import { render } from "@testing-library/react";
import { CartProvider } from "@/context/CartContext";
import { ThemeProvider } from "styled-components";
import { theme } from "@/styles/theme";
import CartSummary from "@/app/cart/components/Cart/CartSummary";

test("CartSummary renders correctly", () => {
  const { asFragment } = render(
    <ThemeProvider theme={theme}>
      <CartProvider>
        <CartSummary />
      </CartProvider>
    </ThemeProvider>
  );

  expect(asFragment()).toMatchSnapshot();
});
