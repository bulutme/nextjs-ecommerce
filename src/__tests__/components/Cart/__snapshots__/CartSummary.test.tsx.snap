// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartSummary renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-322003de-1 fzDnPL"
  >
    <h3
      class="sc-322003de-0 gULtEc"
    >
      Cart Summary
    </h3>
    <div
      class="sc-322003de-2 geIEUV"
    >
      <div
        class="sc-322003de-3 fmzNQp"
      >
        <span
          class="sc-322003de-4 hBTllM"
        >
          Subtotal
        </span>
        <span
          class="sc-322003de-5 dKYRVM"
        >
          $0.00
        </span>
      </div>
      <div
        class="sc-322003de-3 fmzNQp"
      >
        <span
          class="sc-322003de-4 hBTllM"
        >
          Shipping
        </span>
        <span
          class="sc-322003de-5 dKYRVM"
        >
          Free
        </span>
      </div>
      <div
        class="sc-322003de-3 sc-322003de-6 fmzNQp deJeMf"
      >
        <span
          class="sc-322003de-4 hBTllM"
        >
          Total
        </span>
        <span
          class="sc-322003de-5 sc-322003de-7 dKYRVM kGRzLl"
        >
          $0.00
        </span>
      </div>
    </div>
  </div>
  <button
    class="sc-e6c76d24-0 lfwATC"
    color="primary"
  >
    Proceed to checkout
  </button>
</DocumentFragment>
`;
